USB Composite Device Support

Dicklessgreat opened the pull request:
This pull request aims to expand USB functionality by adding support for a composite USB device.

### Motivation

The motivation behind creating a composite USB device is to enable versatile control of my audio device currently under development. In my case:

- USB MIDI Interface: This interface will allow my device to be controlled by DAW software on a host PC, providing integration with standard audio workflows.
- Vendor-Specific Interface (with postcard-rpc support): This secondary interface enables more detailed control of the device via CLI, even if a user does not have a DAW. This addition offers two primary benefits:
    - User Accessibility: Users can control the device using a simple CLI utility, making it accessible even outside of DAW environments.
    - Development & Debugging: From my end, this interface is invaluable for running simple tests from the PC, simplifying debugging during development.

### Details

To implement this, I introduced a closure argument to `new_raw_nusb`, providing flexibility in selecting the interface number from `InterfaceInfo` ([commit ba2ac52](https://github.com/Dicklessgreat/postcard-rpc/commit/ba2ac52741fe835cae0462af1ecec4b622b7c316)). However, I recognize alternative methods might exist, and I'm open to feedback on optimal design approaches.

### Background

As a developer with limited experience in USB standards, I'm aware this approach may not fully align with best practices. If supporting multiple USB classes in a single device proves incompatible with the design goals of this crate, I'm open to revising or abandoning this feature.

jamesmunns commented:
Hey there! I think I'm open to this, would you mind making this a different constructor, rather than modifying the existing one? This would allow us to avoid a breaking change.

Dicklessgreat commented:
Got it! I've made the requested changes. If I understand correctly, I noticed the `HostClient::new_raw_nusb()` constructor is intended specifically for use with "Vendor Specific" interfaces? If that's the case, the current adjustments should align with the expectations. Please let me know if I've misunderstood anything or if further modifications are needed.
Thanks.

jamesmunns reviewed (approved):
I *think* I'm okay with this, with the caveat that if anyone complains that this breaks their setup, I may revert + yank it.

I am also open to more granular/detailed constructors that allow you to more specifically pick the interface (one closure for picking devices, and a second for picking an interface from the matching device?). I just haven't needed it yet. Feedback is definitely welcome, and I'm not a USB expert either!

Thanks!

jamesmunns commented on source/postcard-rpc/src/server/impls/embassy_usb_v0_3.rs:
```diff
@@ -163,7 +163,7 @@ pub mod dispatch_impl {
             driver: D,
             config: Config<'static>,
             tx_buf: &'static mut [u8],
-        ) -> (UsbDevice<'static, D>, WireTxImpl<M, D>, WireRxImpl<D>) {
```
Ah, did you mean to change this? This is also a breaking change (which I want to avoid if possible), and currently breaks CI.

If you want this interface, can you add it as a different (additional) constructor, instead of changing the existing one?

Dicklessgreat replied:
Ah, I completely forgot about this. Thanks for pointing it out!

Dicklessgreat commented:
### `WireStorage` ctor
To avoid introducing a breaking change, I’ve added a new function called `init_without_build()`. This new function is 90% the same as the original `init()` but skips calling `builder.build()` at the end.

The existing `init()` now calls `init_without_build()` internally, followed by `builder.build()`. This approach ensures that there’s no duplication of logic while keeping the original functionality intact.

### `HostClient` init
I’ve added a new function called `try_new_raw_nusb_with_interface()`. It’s essentially a renamed version of what I previously committed, added as a separate function. I thought it would be useful to include this as a precaution(for me) in case e6a0120d8c9c66a1ec7c1455cb70c9aa381ceaed needs to be reverted or yanked later.

This function duplicates the logic of `try_new_raw_nusb()` intentionally, as it makes reverting or yanking easier.

While this duplication is fine for now, an alternative approach could be to modify `try_raw_nusb()` like this (details below). I’m happy to adjust things however is most convenient for you as the maintainer—just let me know!
```
pub fn try_new_raw_nusb<F: FnMut(&DeviceInfo) -> bool>(
    func: F,
    err_uri_path: &str,
    outgoing_depth: usize,
    seq_no_kind: VarSeqKind,
) -> Result<Self, String> {
    try_new_raw_nusb_with_interface(
        func, 
        |i| i.class() == 0xff,
        err_uri_path,
        outgoing_depth,
        seq_no_kind
    )
}
```
